IAU Projekt 1. fáza

Autori: Martin Raffáč, Tibor Vanek


In [15]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib as mat
import matplotlib.pylab as pylab
import statsmodels.api as sm
import statsmodels.stats as sm_stats
import statsmodels.stats.api as sms
import scipy.stats as stats
from matplotlib import pyplot
from collections import Counter

In [16]:
filename1 = "./dataset-003/measurements.csv"
measurements = pd.read_csv(filename1, sep='\t')
measurements.head()




longitude  C2H3NO5       O3       CO  warning        Pb      PAHs    PM2.5  \
0   30.28491  0.01960  8.26626  9.07784      0.0  46.83125   5.53330  5.37316   
1  120.89694  0.04735  8.35627  6.95474      1.0  48.68708  15.29667  7.42089   
2  141.95000  0.04530  7.47394  7.47366      1.0  60.75523  10.47662  8.29646   
3   15.07041  8.64588  8.50563  7.79137      0.0  54.89256   8.17284  8.34238   
4  -94.74049  0.02774  8.01382  9.46041      0.0  61.97648   6.41785  7.84354   

         PRES      TEMP  latitude      NH3      H2CO      SO2      CH4  \
0  1101.06102  33.34540  60.02427  6.23330  54.29137  7.88335  5.85807   
1  1141.23149  20.75105  14.95472  9.32493  53.43339  8.41893  8.32431   
2  1117.39968  21.07022  39.46667  9.15644  31.70985  9.80100  9.24057   
3  1175.94780  11.22310  37.49223  8.47673  58.04300  3.12505  6.73364   
4  1160.95559  -1.44749  32.50070  8.01924  64.31668  9.01725  6.75349   

       PM10      CFCs       NOx  
0   5.68070  28.12016  11.49560  
1   7.01594  82.17695   9.13370  
2   7.72587  19.48463   7.41425  
3  13.35810  55.11303   5.24699  
4   7.07762  50.38845   9.37646

In [17]:
filename2 = "./dataset-003/stations.csv"
stations = pd.read_csv(filename2, sep='\t')
stations.head()

station              revision          QoS  longitude code  \
0  Barra dos Coqueiros            2020-02-07  maintenance  -37.03861   BR   
1          Tyoply Stan  09/13/2018, 00:00:00    excellent   37.49338   RU   
2               Leland            2019-07-09    excellent  -78.04471   US   
3            Hermsdorf           15 Mar 2022    excellent   13.31783   DE   
4           Ulan Bator  12/26/2016, 00:00:00  maintenance  106.88324   MN   

   latitude  
0 -10.90889  
1  55.62047  
2  34.25628  
3  52.61131  
4  47.90771

1. Základný opis dát spolu s ich charakteristikou 

V tejto časti uvedieme:
    ● Počet záznamov, počet atribútov, ich typy,
    ● Pre zvolené významné atribúty ich distribúcie, základné deskriptívne štatistiky a pod.
    ● Párová analýza dát: preskúmajte vzťahy medzi zvolenými dvojicami atribútov.
    ● Párová analýza dát: Identifikujte závislostí medzi dvojicami atribútov (napr. korelácie)
    ● Párová analýza dát: Identifikujte závislosti medzi predikovanou premennou a ostatnými
    premennými (potenciálnymi prediktormi).

dataset measurements.csv obsahuje 12075 záznamov o chemických prvkoch. Počet atribútov v tomto datasete je 18. 

dataset stations.csv obsahuje 1103 záznamov o chemických prvkoch. Počet atribútov v tomto datasete je 6.

Zoznam daných atribútov:

 ●   longitude  číselná hodnota
 ●   C2H3NO5    číselná hodnota
 ●   O3         číselná hodnota
 ●   CO         číselná hodnota
 ●   warning    číselná hodnota
 ●   Pb         číselná hodnota
 ●   PAHs       číselná hodnota 
 ●   PM2.5      číselná hodnota
 ●   PRES       číselná hodnota
 ●   TEMP       číselná hodnota
 ●   latitude   číselná hodnota
 ●   NH3        číselná hodnota
 ●   H2CO       číselná hodnota
 ●   SO2        číselná hodnota
 ●   CH4        číselná hodnota
 ●   PM10       číselná hodnota
 ●   CFCs       číselná hodnota
 ●   NOx        číselná hodnota



In [18]:
measurements.describe()

longitude       C2H3NO5            O3            CO       warning  \
count  12075.000000  12027.000000  12027.000000  12027.000000  12027.000000   
mean      15.192940      0.170331      8.243161      8.226100      0.596491   
std       74.224284      1.055270      0.996025      1.003052      0.490622   
min     -156.474320      0.000000      4.660310      4.464290      0.000000   
25%      -52.355580      0.027100      7.572615      7.541400      0.000000   
50%       11.500000      0.042240      8.243740      8.213650      1.000000   
75%       74.122860      0.098490      8.914240      8.910910      1.000000   
max      178.004170    100.000000     12.364080     11.627400      1.000000   

                 Pb          PAHs         PM2.5          PRES          TEMP  \
count  12027.000000  12026.000000  12027.000000  12075.000000  12027.000000   
mean      50.186935      8.502031      8.229532   1124.800136     14.959069   
std       12.100899      1.931576      1.005640     43.308139      9.934451   
min        0.000000      1.058880      4.218270   1050.001570    -21.492790   
25%       42.104350      7.121487      7.558260   1087.231535      8.249165   
50%       50.260780      8.562045      8.234510   1124.654680     14.956420   
75%       58.312705      9.859282      8.910315   1162.067155     21.660035   
max      100.000000     15.863860     11.944030   1199.998800     54.773100   

           latitude           NH3          H2CO           SO2           CH4  \
count  12075.000000  12027.000000  12027.000000  12027.000000  12026.000000   
mean      28.472685      7.713508     47.566325      9.002650      8.492999   
std       22.845126      1.950890     13.212382      1.672296      1.986720   
min      -54.800000      0.000000      0.000000      1.801440      0.233150   
25%       14.620720      6.370755     38.573510      8.172625      7.148212   
50%       35.899720      7.601500     47.630350      9.221320      8.649600   
75%       44.976390      9.037325     56.478610     10.085010      9.925450   
max       68.798330     15.767780    100.000000     15.619170     14.350810   

               PM10          CFCs           NOx  
count  12026.000000  12025.000000  12026.000000  
mean       7.439091     51.446245      7.460853  
std        1.875983     13.186757      1.727860  
min        0.994780      0.000000      0.596490  
25%        6.165990     42.674950      6.351095  
50%        7.352215     51.426230      7.304430  
75%        8.647887     60.334910      8.525882  
max       15.809190    100.000000     15.055640

In [19]:
len(measurements)

12075

In [20]:
len(measurements.columns)

18

In [21]:
measurements.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12075 entries, 0 to 12074
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   longitude  12075 non-null  float64
 1   C2H3NO5    12027 non-null  float64
 2   O3         12027 non-null  float64
 3   CO         12027 non-null  float64
 4   warning    12027 non-null  float64
 5   Pb         12027 non-null  float64
 6   PAHs       12026 non-null  float64
 7   PM2.5      12027 non-null  float64
 8   PRES       12075 non-null  float64
 9   TEMP       12027 non-null  float64
 10  latitude   12075 non-null  float64
 11  NH3        12027 non-null  float64
 12  H2CO       12027 non-null  float64
 13  SO2        12027 non-null  float64
 14  CH4        12026 non-null  float64
 15  PM10       12026 non-null  float64
 16  CFCs       12025 non-null  float64
 17  NOx        12026 non-null  float64
dtypes: float64(18)
memory usage: 1.7 MB


In [22]:
stations.describe()

longitude     latitude
count  1103.000000  1103.000000
mean     15.178244    28.546164
std      74.259394    22.830913
min    -156.474320   -54.800000
25%     -52.280985    14.625780
50%      11.500000    35.956640
75%      74.216030    44.982700
max     178.004170    68.798330

In [23]:
len(stations)

1103

In [24]:
len(stations.columns)

6

In [25]:
stations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1103 entries, 0 to 1102
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   station    1103 non-null   object 
 1   revision   1103 non-null   object 
 2   QoS        1103 non-null   object 
 3   longitude  1103 non-null   float64
 4   code       1103 non-null   object 
 5   latitude   1103 non-null   float64
dtypes: float64(2), object(4)
memory usage: 51.8+ KB


2. Identifikácia problémov v dátach s prvotným riešením

nevhodná štruktúra:
duplicitné záznamy: mam
nejednotné formáty:
chýbajúce hodnoty: mam
vychýlené hodnoty:
iné problémy:

In [26]:
# Vyhľadanie duplicitných záznamov

duplicate = measurements[measurements.duplicated()]
measurements = measurements.drop_duplicates()
print("Počet záznamov v datasete: "+str(len(measurements)))
print("V datasete sa nachádza: " + str(len(duplicate)) + " duplikátov.")


Počet záznamov v datasete: 11897
V datasete sa nachádza: 178 duplikátov.


Pomocou týchto príkazov sme zistili či sa v datasete nachádzajú duplicitné záznamy, program detegoval 178 duplikátnych záznamov.
Tieto duplikáty sme z datasetu vymazali aby nespôsobovali chyby pri neskoršej práci s dátami.


In [83]:
measurements.isnull().sum()

longitude     0
C2H3NO5      48
O3           48
CO           48
warning      48
Pb           48
PAHs         48
PM2.5        48
PRES          0
TEMP         48
latitude      0
NH3          48
H2CO         48
SO2          48
CH4          48
PM10         48
CFCs         48
NOx          48
dtype: int64

Z tejto štatistiky vyplíva, že viacero atribútov má nulové hodnoty. Polia s nulovou hodnotou sme nahradili priemerom z dôvodu aby sme neprišli o dáta.

In [85]:
C2H3NO5_mean = measurements['C2H3NO5'].mean()
measurements['C2H3NO5'] = measurements['C2H3NO5'].fillna(C2H3NO5_mean)
O3_mean = measurements['O3'].mean()
measurements['O3'] = measurements['O3'].fillna(O3_mean)
CO_mean = measurements['CO'].mean()
measurements['CO'] = measurements['CO'].fillna(CO_mean)
Pb_mean = measurements['Pb'].mean()
measurements['Pb'] = measurements['Pb'].fillna(Pb_mean)
PAHs_mean = measurements['PAHs'].mean()
measurements['PAHs'] = measurements['PAHs'].fillna(PAHs_mean)
PM25_mean = measurements['PM2.5'].mean()
measurements['PM2.5'] = measurements['PM2.5'].fillna(PM25_mean)
TEMP_mean = measurements['TEMP'].mean()
measurements['TEMP'] = measurements['TEMP'].fillna(TEMP_mean)
NH3_mean = measurements['NH3'].mean()
measurements['NH3'] = measurements['NH3'].fillna(NH3_mean)
H2CO_mean = measurements['H2CO'].mean()
measurements['H2CO'] = measurements['H2CO'].fillna(H2CO_mean)
SO2_mean = measurements['SO2'].mean()
measurements['SO2'] = measurements['SO2'].fillna(SO2_mean)
CH4_mean = measurements['CH4'].mean()
measurements['CH4'] = measurements['CH4'].fillna(CH4_mean)
PM10_mean = measurements['PM10'].mean()
measurements['PM10'] = measurements['PM10'].fillna(PM10_mean)
CFCs_mean = measurements['CFCs'].mean()
measurements['CFCs'] = measurements['CFCs'].fillna(CFCs_mean)
NOx_mean = measurements['NOx'].mean()
measurements['NOx'] = measurements['NOx'].fillna(NOx_mean)

Keďže atribút warning sme nemohli nahradiť priemernou hodnotou lebo v datasete sa viskitoval len s hodnotami 1 alebo 0
rozhodli sme sa vypočítať priemer a zaokruhliť ho na najbližšie celé číslo a následne toto číslo doplniť do prázdnych polí.

In [88]:
warning_avg = measurements['warning'].mean()

if warning_avg >= 0.5:
    warning_avg = 1
else:
    warning_avg = 0

measurements['warning'] = measurements['warning'].fillna(warning_avg)

In [89]:
measurements.isnull().sum()

longitude    0
C2H3NO5      0
O3           0
CO           0
warning      0
Pb           0
PAHs         0
PM2.5        0
PRES         0
TEMP         0
latitude     0
NH3          0
H2CO         0
SO2          0
CH4          0
PM10         0
CFCs         0
NOx          0
dtype: int64

Atribúty z tohoto datasetu measurements.csv majú číslené hodnoty. Atribúty longitude a latitude budeme potrebovať na spojenie s datasetom stations.csv preože tieto atribúty sú spoločné. 

In [90]:
measurements.isnum().count()

AttributeError: 'DataFrame' object has no attribute 'isnum'